### DISCUSSION TITLE
Custom cronjobs

### DISCUSSION THREAD

**Discussion by Zoey2936** [nextcloud/all-in-one#2912] [Ideas] [answered]
For example, by triggering a script inside one of the nextcloud volumes

**Comment by szaimen** [comment]
Why would you need this?

**Reply by Zoey2936** [reply]
https://github.com/nextcloud/all-in-one/discussions/2810 for example

**Reply by szaimen** [reply]
Do you have another example?

**Reply by Zoey2936** [reply]
at the moment not, i will search something in three weeks

**Reply by szaimen** [reply]
Okay

**Reply by Zoey2936** [reply]
https://github.com/nextcloud/all-in-one/discussions/1581 that would also be a usecase

**Reply by szaimen** [reply]
> #1581 that would also be a usecase

could be a follow-up for https://github.com/nextcloud/all-in-one/discussions/1581 then

**Reply by Zoey2936** [reply]
what do you mean?

**Reply by szaimen** [reply]
I added it to https://github.com/nextcloud/all-in-one/discussions/1581 (there is a follow-up section)

**Reply by Zoey2936** [reply]
i see

**Comment by szaimen** [comment, marked as answer]
I thought about this a bit more and currently there are no plans to add support for this as you can use host cronjobs for this.

**Reply by gggeek** [reply]
There are containers which are dedicated to running a (centralized) job scheduler, able to run commands on all the containers in the stack.

Maybe we could think about adding one of those to AIO?

Advantages:
- a single process instead of one-per-container
- could allow community containers to define their own scheduled jobs, that we forward to the scheduler container
- does not rely on running cron within containers (which is horrible to troubleshoot)

**Reply by gggeek** [reply]
Related: #2829

**Reply by szaimen** [reply]
Yes, this probably is doable as a community container. PRs are welcome!

**Reply by gggeek** [reply]
@szaimen I thought more about this.

Initially, I was going to propose using Ofelia, a widely-known tool, generally delivered as a container, which can run jobs on other containers via mounting the docker socket.

Then I realised that:
- we should have as few containers as possible within the AIO that mount the docker socket, as each has basically root access to the host! (a mitigation would be to only allow access through a semantically-aware docket-socket-proxy, but that is not yet the case (it is tracked in #1713, see comment https://github.com/nextcloud/all-in-one/discussions/1713#discussioncomment-16974508))
- atm, the container which runs most cron tasks is the MC (in fact, it relies on a kind of hackish series of shell-based scripts to simulate a proper scheduler)
- the MC already mounts the docker socket

Which makes the MC the ideal candidate for running a proper scheduler daemon :-)

What I propose is thus:
1. select a candidate daemon to run as job scheduler within the MC
2. have it built in, started at start.sh time
3. replace the current set of cronjobbish shell scripts in the MC with proper cron job defs
4. allow the container definition in containers.json to specify jobs which should be run on the other containers

About point 1:
- it could be a long-running php script. I love (ab)using php for everything, but having a long-running php process is a recipe for resource leaks. Also, it might burn too much cpu on small installs.
- similarly, using a never-stopping shell-script seems not to give us enough flexibility
- we could still use Ofelia, which is available as standalone binary
- or go for old-school cron, but cron is a beast to troubleshoot within containers (it writes to syslog instead of a log file, it has a weird way of acting properly when jobs have stderr/stdout/error-exit-codes)
- I m a great fan of using the crontab format for scheduling tasks
- which leads me to think we could explore crond alternatives written in rust? 

About point 4: that would be done by the same php code in charge of start/stopping the containers. At child container start, add to the local job definitions pool the ones of the child container. At stop, remove them.

**Reply by szaimen** [reply]
Generally the idea would make sense but we need to also consider deployment options like the manual-install and the helm chart where the mastercontainer is not available

**Reply by gggeek** [reply]
Fair point.

So far, there are no job schedulers available at all within the stack, so adding one within the MC should not break anything that is already running / in place.
But if the non-MC containers come to rely on having jobs scheduled, we should provide an alternative to the MC being available for running them.

How to achieve that: off the top of my head I can propose this (and feel free to come up with better ideas, it does feel sub-optimal even as I type):

- manual install: when generating the yaml file, add to it the definition of one extra container which will run the jobs. When doing that, we should parse the job defs found in containers.json and add them to compose.yaml. Nothing that jq/yq can not do, even though it might look like dirty code (this might be in fact a good reason to use Ofelia as scheduler, as it is easy to bundle it as standalone container. Or maybe just build a container running crond and call it a day? But then we have to prepend to every job spec something like `docker exec -t $container `)

- helm charts: I am out of my depth here, as my experience with the K8s ecosystem is limited. I think there is a concept of cron jobs that can be defined, but those spin up new pods instead of running inside existing ones. Which we might possibly adopt as strategy even inside the MC-managed-stack, but it opens up an altogether different can of worms, I think, as it means checking how the AIO containers react to being started up with a different command that start.sh...

**Reply by gggeek** [reply]
As Jake would say: "I have seen the light!"
Taking a break and having some food (and coffee!) always helps with lateral thinking.
Either that, or I had too much wine :-P

The maximum-simplicity-design solution is: always run the MC container, even when in manual-install mode and in K8s contexts. But, when in those contexts, do not start its web GUI, nor its upgrade/backup crons. 

This way the MC can still deliver the different services which can be considered "core" to the platform:
- acting as a docker-socket proxy for other containers (cf #3230)
- acting as a job scheduler for other containers (cf #2912)
- acting as an SSO server for other containers (cf #7686)
- allow a cli user on the host to use the MC cli tool to run one-off tasks, such as unscheduled backups (cf #8156, #8158)
- act as an HARP proxy ?

**Reply by szaimen** [reply]
I fear the mastercontainer will not work with the helm chart as the docker socket is technically not accessible there afaik

**Reply by gggeek** [reply]
Ok, so in the helm chart just set up the MC not to act as docker-socket proxy.
It starts to feel that we are stretching the MC a bit in too many directions, but I would say we are still a few inches on this side of the "building a frankenmonster" precipice.
